### PR TITLE
Don't stop checking if the `HealthNode` persistent task is present

### LINKED ISSUE #98926: Health API reports disk information `Unknown` status / `No disk usage data` symptom
### Elasticsearch Version

8.9.1

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

N.A

### Problem Description

Elasticsearch Service (ESS) users have observed the below error message when accessing the `Health` API of a deployment:

```
Error: Cannot read properties of undefined (reading 'replace")
at https://cloud.elastic.co/app.070493148d8f8558ab6.js:2:3103121
at U1 (https://cloud.elastic.co/app.070493148d88558ab6.js:2:3103141) 
at a (https://cloud.elastic.co/vendor.3ccc0c4579678245a4e.js:2:5829841) 
at $ (https://cloud.elastic.co/vendor.3ccc0c4579678245a4e.js:2:5881966) 
at Tl (https: //cloud.elastic.co/vendor.3ccc@c4579678245a4e. js:2:5869244) 
at El (https: //cloud.elastic.co/vendor.3ccc@c4579678245a4e.j:2:5869172) 
at _1 (https://cloud.elastic.co/vendor.3ccc0c4579678245a4e.js:2:5869035) 
at yl (https: //cloud.elastic.co/vendor.3ccc@c4579678245a4e. js:2:5866022)
at https://cloud.elastic.co/vendor.3ccc0c4579678245a4e.js:2:5815760
at t.unstable_runWithPriority (https://cloud.elastic.co/vendor.3ccc@c457967e8245a4e. js:2:6649725)
at Go (https: //cloud.elastic.co/vendor.3ccc@c457967e8245a4e. js:2:5815537) 
at Xo (https: //cloud.elastic.co/vendor.3ccc@c4579678245a4e. js:2:5815705) 
at $o (https: //cloud.elastic.co/vendor.3ccc0c457967e8245a4e.js:2:5815640) 
at Ml (https: //cloud.elastic.co/vendor. 3ccc0c4579678245a4e. js:2:5866343)
at Object.notify (https: //cloud.elastic.co/vendor.3ccc@c457967e8245a4e. js:2:5977078)
at Object.notifyNestedSubs (https://cloud.elastic.co/vendor.3ccc@c457967e8245a4e. js:2:5977491)
```

Further investigation revealed that the Cloud UI is unable to gracefully handle the following `disk` information returned by the Health API

```
      "disk" : {
        "status" : "Unknown",
        "symptom" : "No disk usage data."
      },
```

While the above error can be better handled in ESS, it is unclear why the Health API reports such disk status / symptom. This requires further investigation.

### Workaround

Users can disable and enable again the health node in Elasticsearch. These are the required APIs:

- Disable health node:

```
PUT _cluster/settings
{
  "persistent": {
    "health.node.enabled": false
  }
}
```

- Enable health node:

```
PUT _cluster/settings
{
  "persistent": {
    "health.node.enabled": true
  }
}
```

Once the two APIs are executed, observe that the Health page is accessible again.

### Steps to Reproduce

- In ESS, create a deployment:
  - 2GB RAM x 2AZ (hot tier)
  - Version 7.17.6
- Wait for the deployment to be created, take a snapshot:
   - `POST _slm/policy/cloud-snapshot-policy/_execute`
- Create another deployment:
  - Same topology (same region, 2GB RAM x 2AZ (hot tier))
  - Version 8.9.1
  - Restore snapshot from the 7.17.6 deployment (by toggling the `Restore snapshot data` when creating the deployment)
- Wait for the deployment version 8.9.1 to be created, access the `Health` page and observe the above error message.
- Run the Health API:

```
{
  "status": "unknown",
  "cluster_name": "xxx",
  "indicators": {
    "master_is_stable": {
      "status": "green",
      "symptom": "The cluster has a stable master node",
      "details": {
        "current_master": {
          "node_id": "4XoNLHsXR3aiO4kqb_x__g",
          "name": "instance-0000000001"
        },
        "recent_masters": [
          {
            "node_id": "YE16I4OrQP60r0A9hGYMhg",
            "name": "instance-0000000000"
          },
          {
            "node_id": "4XoNLHsXR3aiO4kqb_x__g",
            "name": "instance-0000000001"
          }
        ]
      }
    },
    "repository_integrity": {
      "status": "green",
      "symptom": "No corrupted snapshot repositories.",
      "details": {
        "total_repositories": 2
      }
    },
    "shards_availability": {
      "status": "green",
      "symptom": "This cluster has all shards available.",
      "details": {
        "started_primaries": 27,
        "unassigned_primaries": 0,
        "initializing_replicas": 0,
        "started_replicas": 27,
        "initializing_primaries": 0,
        "restarting_replicas": 0,
        "restarting_primaries": 0,
        "unassigned_replicas": 0,
        "creating_primaries": 0
      }
    },
    "disk": {
      "status": "unknown",
      "symptom": "No disk usage data."
    },
    "shards_capacity": {
      "status": "green",
      "symptom": "The cluster has enough room to add new shards.",
      "details": {
        "data": {
          "max_shards_in_cluster": 2000
        },
        "frozen": {
          "max_shards_in_cluster": 0
        }
      }
    },
    "ilm": {
      "status": "green",
      "symptom": "Index Lifecycle Management is running",
      "details": {
        "policies": 29,
        "stagnating_indices": 0,
        "ilm_status": "RUNNING"
      }
    },
    "slm": {
      "status": "green",
      "symptom": "Snapshot Lifecycle Management is running",
      "details": {
        "slm_status": "RUNNING",
        "policies": 1
      }
    }
  }
}
```

### Logs (if relevant)

After setting the `logger.org.elasticsearch.health.node` to `DEBUG` and restarting the current master (instance 0 in our example), the below DEBUG logs were observed:

```
  | 28 Aug, 2023 @ 09:33:08.070 | [instance-0000000000] Resetting the health monitoring because the master node changed, current health node is null.

  | 28 Aug, 2023 @ 09:30:55.653 | [instance-0000000001] Resetting the health monitoring because the master node changed, current health node is null.

  | 28 Aug, 2023 @ 09:30:55.145 | [tiebreaker-0000000002] Resetting the health monitoring because the master node changed, current health node is null. 

  | 28 Aug, 2023 @ 09:30:54.647 | [instance-0000000001] Resetting the health monitoring because the master node changed, current health node is null

  | 28 Aug, 2023 @ 09:30:54.630 | [tiebreaker-0000000002] Resetting the health monitoring because the master node changed, current health node is null
```

### PR DESCRIPTION
**Pull request by nielsbauman**
We assumed that once the `HealthNode` persistent task is registered, we won't need to register it again. However, when, for instance, we restore from a snapshot (including cluster state) that was created in version <= 8.4.3, that task doesn't exist yet, which will result in the task being removed after the restore. By keeping the listener active, we will re-add the task after such a restore (or any other potential situation where the task might get deleted).

Fixes #98926

### CODE REVIEW

**Review by gmarouli — COMMENTED** [review]
What do you think about adding a test for this?

**Comment by nielsbauman** [comment]
@gmarouli what kind of test did you have in mind? I thought about testing as well but couldn't really come up with a good way to test this (but forgot to write that in the PR). I think that
1. unit testing is going to be a bit hard here, as there wasn't really anything wrong with the "unit" (i.e. the `HealthNodeTaskExecutor` class/`runTask` method), and testing that `removeListener` is not called doesn't make a lot of sense to me.
2. adding a test [here](https://github.com/nielsbauman/elasticsearch/blob/6066d1d46ea9cbe99904361a30af3ae4203b4dfa/qa/repository-multi-version/src/test/java/org/elasticsearch/upgrades/MultiVersionRepositoryAccessIT.java#L48-L48) (David showed me how) would allow us to test this, but it feels a bit off testing this way, as the problem isn't technically related to the snapshot restore process; it's related to us removing the listener.
3. the ideal way would be to have an integration test where we 1. assess health node is running, 2. somehow remove the health task, and 3. verify that the health task got started again. Do you know if step 2 here is (easily) possible?

Or perhaps you had another way of testing in mind?

**Comment by gmarouli** [comment]
> @gmarouli what kind of test did you have in mind? I thought about testing as well but couldn't really come up with a good way to test this (but forgot to write that in the PR). I think that
> 
> 1. unit testing is going to be a bit hard here, as there wasn't really anything wrong with the "unit" (i.e. the `HealthNodeTaskExecutor` class/`runTask` method), and testing that `removeListener` is not called doesn't make a lot of sense to me.
> 2. adding a test [here](https://github.com/nielsbauman/elasticsearch/blob/6066d1d46ea9cbe99904361a30af3ae4203b4dfa/qa/repository-multi-version/src/test/java/org/elasticsearch/upgrades/MultiVersionRepositoryAccessIT.java#L48-L48) (David showed me how) would allow us to test this, but it feels a bit off testing this way, as the problem isn't technically related to the snapshot restore process; it's related to us removing the listener.
> 3. the ideal way would be to have an integration test where we 1. assess health node is running, 2. somehow remove the health task, and 3. verify that the health task got started again. Do you know if step 2 here is (easily) possible?
> 
> Or perhaps you had another way of testing in mind?

Hm, fair point. I agree with your analysis above. The last one sounds great but I am not aware of how to delete the task. What about changing the current unit test:
https://github.com/elastic/elasticsearch/blob/2cec43d0a8e49ca3debde64084803799cab78313/server/src/test/java/org/elasticsearch/health/node/selection/HealthNodeExecutorTests.java#L94-L109

to this:
```
public void testTaskCreation() throws Exception {
        HealthNodeTaskExecutor.create(clusterService, persistentTasksService, featureService, settings, clusterSettings);
        clusterService.getClusterApplierService().onNewClusterState("initialization", this::initialState, ActionListener.noop());
        // Ensure that if the task is gone, it will be recreated.
        clusterService.getClusterApplierService().onNewClusterState("initialization", this::initialState, ActionListener.noop());
        assertBusy(
            () -> verify(persistentTasksService, times(2)).sendStartRequest(
                eq("health-node"),
                eq("health-node"),
                eq(new HealthNodeTaskParams()),
                any()
            )
        );
    }
```

PS: If it's not much trouble can you rename `HealthNodeExecutorTests` to `HealthNodeTaskExecutorTests`. This is probably my mistake and it's causing IDE navigation to not detect the test.

**Comment by nielsbauman** [comment]
@gmarouli I quite like that solution, thanks a lot for the suggestion! And will rename the test class 👍🏻 .

File: server/src/test/java/org/elasticsearch/health/node/LocalHealthMonitorTests.java [hunk 1]
```diff
@@ -63,7 +62,7 @@ public class LocalHealthMonitorTests extends ESTestCase {
 
     @BeforeClass
     public static void setUpThreadPool() {
-        threadPool = new TestThreadPool(HealthNodeExecutorTests.class.getSimpleName());
+        threadPool = new TestThreadPool(LocalHealthMonitorTests.class.getSimpleName());
```
Comment (gmarouli) [inline server/src/test/java/org/elasticsearch/health/node/LocalHealthMonitorTests.java:65]: 👏

**Review by gmarouli — APPROVED** [review]
LGTM, thank you for fixing this @nielsbauman !

File: server/src/test/java/org/elasticsearch/health/node/LocalHealthMonitorTests.java (hunk 1 above)
Comment (nielsbauman) [reply server/src/test/java/org/elasticsearch/health/node/LocalHealthMonitorTests.java:65]: Credits to Intellij refactoring functionality 😉.

**Pull request merged**